### PR TITLE
[tests] gpt2 + `CausalLMModelTester`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR is an example of a non-super-standard decoder-only model with `CausalLMModelTester`: with a few extra definitions, we can reuse most of the improved testing logic.

👉 If this pattern is approved, I'll open a separate PR for other decoder-only models 🤗

### CODE REVIEW

File: tests/causal_lm_tester.py [hunk 1]
```diff
@@ -316,6 +316,27 @@ def test_token_classification_model(self):
             (self.model_tester.batch_size, self.model_tester.seq_length, self.model_tester.num_labels),
         )
 
+    def test_question_answering_model(self):
```
Comment (gante) [inline tests/causal_lm_tester.py:319]: `CausalLMModelTester` defines a `question_answering_class`, but it was missing the standard test.

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 2]
```diff
@@ -48,443 +45,78 @@
...
+class GPT2ModelTester(CausalLMModelTester):
+    if is_torch_available():
+        config_class = GPT2Config
+        base_model_class = GPT2Model
+        causal_lm_class = GPT2LMHeadModel
+        sequence_classification_class = GPT2ForSequenceClassification
+        token_classification_class = GPT2ForTokenClassification
+        question_answering_class = GPT2ForQuestionAnswering
+
     def __init__(
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:57]: overloads `__init__` with non-standard settings:
- `use_token_type_ids=True`, because this model uses these tokens 
- `num_choices=4` for a model-specific class

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 3]
```diff
@@ -48,443 +45,78 @@
...
-        self.bos_token_id = vocab_size - 1
-        self.eos_token_id = vocab_size - 1
-        self.pad_token_id = vocab_size - 1
-
-    def get_large_model_config(self):
-        return GPT2Config.from_pretrained("openai-community/gpt2")
 
     def prepare_config_and_inputs(
-        self, gradient_checkpointing=False, scale_attn_by_inverse_layer_idx=False, reorder_and_upcast_attn=False
+        self, full_inputs=False, scale_attn_by_inverse_layer_idx=False, reorder_and_upcast_attn=False
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py]: - `full_inputs`: the model-specific class (`GPT2DoubleHeadsModel`) needs special inputs
- other flags: gpt2-specific config flags that are set and tested below

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 4]
```diff
@@ -48,443 +45,78 @@
...
-
-        inputs_dict = {
-            "input_ids": input_ids,
-            "token_type_ids": token_type_ids,
-            "head_mask": head_mask,
-        }
-
+        config_and_inputs = self.prepare_config_and_inputs(full_inputs=True)
+        config, input_ids, _, head_mask, token_type_ids, _, _, _, _ = config_and_inputs
+        inputs_dict = {"input_ids": input_ids, "token_type_ids": token_type_ids, "head_mask": head_mask}
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:118]: prepares `token_type_ids` as part of its common test inputs

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 5]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-    def setUp(self):
-        self.model_tester = GPT2ModelTester(self)
-        self.config_tester = ConfigTester(self, config_class=GPT2Config, n_embd=37)
-
-    def tearDown(self):
-        super().tearDown()
-        # clean-up as much as possible GPU memory occupied by PyTorch
-        cleanup(torch_device)
-
-    def test_config(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:549]: tested in `CausalLMModelTester`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 6]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        self.config_tester.run_common_tests()
-
-    def test_gpt2_model(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:552]: tested in `CausalLMModelTester`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 7]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_model(*config_and_inputs)
-
-    def test_gpt2_model_past(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:556]: redundant with `GenerationMixin` tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 8]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_model_past(*config_and_inputs)
-
-    def test_gpt2_model_att_mask_past(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:560]: redundant with `GenerationMixin` tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 9]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_model_attention_mask_past(*config_and_inputs)
-
-    def test_gpt2_model_past_large_inputs(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:564]: redundant with `GenerationMixin` tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 10]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_model_past_large_inputs(*config_and_inputs)
-
-    def test_gpt2_lm_head_model(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:568]: redundant with `GenerationMixin` tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 11]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_lm_head_model(*config_and_inputs)
-
     def test_gpt2_double_lm_head_model(self):
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_double_lm_head_model(*config_and_inputs)
-
-    def test_gpt2_question_answering_model(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:576]: tested in `CausalLMModelTester`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 12]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_for_question_answering(*config_and_inputs)
-
-    def test_gpt2_sequence_classification_model(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:580]: tested in `CausalLMModelTester`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 13]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_for_sequence_classification(*config_and_inputs)
-
-    def test_gpt2_token_classification_model(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:584]: tested in `CausalLMModelTester`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 14]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_for_token_classification(*config_and_inputs)
-
-    def test_gpt2_gradient_checkpointing(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:588]: we have much better gradient checkpoints in the common tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 15]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_forward_and_backwards(*config_and_inputs, gradient_checkpointing=True)
-
-    def test_gpt2_scale_attn_by_inverse_layer_idx(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:592]: (this was kept, see below)

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 16]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs(scale_attn_by_inverse_layer_idx=True)
-        self.model_tester.create_and_check_forward_and_backwards(*config_and_inputs)
-
-    def test_gpt2_reorder_and_upcast_attn(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:596]: (this was kept, see below)

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 17]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs(reorder_and_upcast_attn=True)
-        self.model_tester.create_and_check_forward_and_backwards(*config_and_inputs)
-
-    def test_gpt2_weight_initialization(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:600]: we have better generalist initialization tests

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 18]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        config_and_inputs = self.model_tester.prepare_config_and_inputs()
-        self.model_tester.create_and_check_gpt2_weight_initialization(*config_and_inputs)
-
-    def test_cached_forward_with_and_without_attention_mask(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:604]: redundant with `GenerationMixin`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 19]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        pass
-
-    @unittest.skip(
-        reason="This architecture seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
-    )
-    def test_training_gradient_checkpointing_use_reentrant_false(self):
-        pass
-
-    @slow
-    def test_batch_generation(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:627]: this is an integration test, moved below

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 20]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        self.assertListEqual(expected_output_sentence, batch_out_sentence)
-        self.assertTrue(batch_out_sentence_tt != batch_out_sentence)  # token_type_ids should change output
-        self.assertListEqual(expected_output_sentence, [non_padded_sentence, padded_sentence])
+    def test_gpt2_scale_attn_by_inverse_layer_idx(self):
+        # extra test: model-specific flag
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(scale_attn_by_inverse_layer_idx=True)
+        config, input_ids, token_type_ids, _, _, _, _ = config_and_inputs
 
-    @slow
-    def test_batch_generation_2heads(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:688]: this is an integration test, moved below

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 21]
```diff
@@ -537,220 +170,91 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
...
-        self.assertListEqual(expected_output_sentence, [non_padded_sentence, padded_sentence])
+    def test_training_gradient_checkpointing_use_reentrant(self):
+        # overwritten: GPT2DoubleHeadsModel fails this test, non-standard class
+        self.original_all_model_classes = self.all_model_classes
+        self.all_model_classes = (cls for cls in self.all_model_classes if cls.__name__ != "GPT2DoubleHeadsModel")
+        super().test_training_gradient_checkpointing_use_reentrant()
+        self.all_model_classes = self.original_all_model_classes
 
-    @slow
-    def test_model_from_pretrained(self):
```
Comment (gante) [inline tests/models/gpt2/test_modeling_gpt2.py:750]: implicitly tested in all integration tests

**Review by Cyrilvallez — COMMENTED** [review]
Looks super nice! Much simpler to follow. Even though it's a very old model, we need to keep very good coverage of it as it's still an important one, but I see you checked every individual test to make sure it was redundant or not 🤗❤️
Looks like it had a blast radius on other test files though!

**Comment by gante** [comment]
@Cyrilvallez all clear now 🤗 (old composite models needed minor tweaks)

I've also added brief comments on all extra functions in `GPT2ModelTester` and `GPT2ModelTest` explaining why they exist -- hopefully, this will set the pattern of documenting all overwrites/extra functions on top of `CausalLMModelTest`

(and yes, I've double-checked all deleted tests to ensure we test the properties they were testing!)

**Review by Cyrilvallez — APPROVED** [review]
LGTM! Thanks for the cleanup! 🤗
